### DISCUSSION TITLE
Error when running in debug mode DGKernel

### DISCUSSION THREAD

**Discussion by ambehnam** [idaholab/moose#21897] [Q&A General] [answered]
Hi everyone,

I have an input file of a 2 dimensional model including TotalLagrangianStressDivergence kernel and my DGkernel. When I run it in dubug mode, I get the following error:

```
Assertion `_current_subdomain_id == _current_elem->subdomain_id()' failed
current subdomain has been set incorrectly
at /home/abehnam/Repositories/moose6/framework/src/base/Assembly.C, line 1969 
```

But, when I run it in opt mode, the analysis converges quickly. I was wondering what may cause this error in debug mode?

It is noticeable that, in my DGKernel, I pass "const Assembly&  _assembly" as input argument to a utility class that computes some penalty tensors. However, even in the cases where that part of the code is not executed, I get the same error in debug mode. 

Thanks a lot,
Amir

The input file is pasted below:
```
[Mesh]
  [./gen]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 2
    xmax = 2
    ny = 1
    ymax = 1
    elem_type = QUAD4
  [../]
  [gen_mod]
    type = MoveNodeGenerator
    input = gen
    node_id = '2'
    new_position = '1.05 1 0'
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y'
  large_kinematics = false
[]

[Variables]
  [disp_x]
  []
  [disp_y]
  []
[]

[Modules]
  [TensorMechanics]
    [Master]
      [all]
        strain = FINITE
        new_system = true
        add_variables = true
        formulation = TOTAL
        volumetric_locking_correction = false
        generate_output = 'cauchy_stress_xx cauchy_stress_yy cauchy_stress_xy '
                          'mechanical_strain_xx mechanical_strain_yy mechanical_strain_xy'                 
      []
    []
  []
[] 

[DGKernels]
  [./vmns_x]
    type = VMNSDGKernelSimple
    variable = disp_x
    component = 0
    nis_flag = -1
    trac_type = 'simple'
  [../]
  [./vmns_y]
    type = VMNSDGKernelSimple
    variable = disp_y
    component = 1
    nis_flag = -1
    trac_type = 'simple'
  [../]
[]

[Materials]
  [elastic_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 190000.0
    poissons_ratio = 0.3
  []
  [compute_stress]
    type = ComputeLagrangianLinearElasticStress
  []
[]

[BCs]
  [left]
    type = DirichletBC
    preset = true
    variable = disp_x
    boundary = left
    value = 0.0
  []
  [bottom]
    type = DirichletBC
    preset = true
    variable = disp_y
    boundary = bottom
    value = 0.0
  []
  [right]
    type = DirichletBC
    preset = true
    variable = disp_x
    boundary = right
    value = 0.1
  []
[]

[Executioner]
  type = Steady
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = 'lu superlu_dist'
  solve_type = NEWTON
[]

[Outputs]
  exodus = true
[]

```

**Comment by GiudGiud** [comment, marked as answer]
Hello

Passing _assembly to a utility does seem a little too deep into the APIs. What attributes are you needing from this?

Guillaume

**Reply by ambehnam** [reply]
Hi Guillaume. The attributes I need are as follows: 
    _current_elem(_assembly.elem()),
    _neighbor_elem(_assembly.neighbor()),
    _current_side(_assembly.side()),
    _neighbor_side(_assembly.neighborSide())

Previously I was using InterfaceMaterial and InterfaceKernel and these attributes were available in InterfaceMaterial. In DGKernel, the first 3 are available in  DGKernelBase, but the last one is not available. So I  used _assembly as input to access all of them.

Thanks,
Amir

**Reply by GiudGiud** [reply]
I dont understand how you can hit this assert when there is no other block ID on the domain. 
Can you add a print statement to see what the two values are?
